todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/domain/logic.kt

```diff
@@ -0,0 +1,38 @@
+@file:JvmName("TaskDetailLogic")
+
+package com.example.android.architecture.blueprints.todoapp.taskdetail.domain
+
+import com.example.android.architecture.blueprints.todoapp.data.Task
+import com.spotify.mobius.Effects.effects
+import com.spotify.mobius.Next
+import com.spotify.mobius.Next.*
+
+fun update(task: Task, event: TaskDetailEvent): Next<Task, TaskDetailEffect> =
+        when (event) {
+            TaskDetailEvent.DeleteTaskRequested -> dispatch(effects(TaskDetailEffect.DeleteTask(task)))
+            TaskDetailEvent.CompleteTaskRequested -> onCompleteTaskRequested(task)
+            TaskDetailEvent.ActivateTaskRequested -> onActivateTaskRequested(task)
+            TaskDetailEvent.EditTaskRequested -> dispatch(effects(TaskDetailEffect.OpenTaskEditor(task)))
+            TaskDetailEvent.TaskDeleted -> dispatch(effects(TaskDetailEffect.Exit))
+            TaskDetailEvent.TaskMarkedComplete -> dispatch(effects(TaskDetailEffect.NotifyTaskMarkedComplete))
+            TaskDetailEvent.TaskMarkedActive -> dispatch(effects(TaskDetailEffect.NotifyTaskMarkedActive))
+            TaskDetailEvent.TaskSaveFailed -> noChange()
+            TaskDetailEvent.TaskDeletionFailed -> noChange()
+        }
+
+private fun onActivateTaskRequested(task: Task): Next<Task, TaskDetailEffect> {
+    if (!task.details.completed) {
+        return noChange<Task, TaskDetailEffect>()
+    }
+
+    val activatedTask = task.activate()
+    return next(activatedTask, effects(TaskDetailEffect.SaveTask(activatedTask)))
+}
+
+private fun onCompleteTaskRequested(task: Task): Next<Task, TaskDetailEffect> {
```

Comment:
Writing this as a single-expression function would be more idiomatic. Same goes for `onActivateTaskRequested`

Reply:
You mean the following?
```
if (task.details.completed) noChange()
else next(task.complete(), effects(SaveTask(task.complete()))
```

Reply:
could probably also be written as
```
if (task.details.completed) noChange()
else task.complete().run { next(this, saveTask(this)) }
```
to avoid calling `complete()` twice

Reply:
I went with @jeppes implementation.